### ISSUE TITLE
[@typescript-eslint/consistent-type-assertions] False positive on specific case

### ISSUE THREAD

**Issue by Alecell** [typescript-eslint/typescript-eslint#2366]
**Repro**
I was trying to use the @typescript-eslint/consistent-type-assertions rule, I was believing that this rule was broken since it was the only rule that I don't get error on editor, but after several hour searching for this I realize that this rule is working, but on that case bellow it don't show any error either on console with `eslint .` or on editor. 

Case: 
```ts
import { action } from 'typesafe-actions';
import { CampaignsRequestTypes, CampaignsState } from './types';

export class CampaignActions {
  static readonly loadList = <const>{
    request: () => action(CampaignsRequestTypes.LOAD_REQUEST),
    success: (data: CampaignsState['data'] = {}) => action(CampaignsRequestTypes.LOAD_SUCCESS, data),
    failure: () => action(CampaignsRequestTypes.LOAD_FAILURE),
  };
}

```

.eslintrc.js
```js
module.exports = {
  env: {
    browser: true,
    es6: true,
    jest: true,
  },
  extends: [
    'react-app',
    'airbnb',
    'plugin:@typescript-eslint/recommended',
  ],
  globals: {
    Atomics: 'readonly',
    SharedArrayBuffer: 'readonly',
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 2018,
    sourceType: 'module',
  },
  plugins: ['react', 'import', 'jsx-a11y', 'es'],
  rules: {
    semi: 'off',
    'multiline-ternary': ['error', 'never'],
    'no-console': 'off',

    'react/jsx-filename-extension': ['error',
      { extensions: ['.tsx'] },
    ],

    'import/prefer-default-export': 'off',
    'import/no-namespace': ['error'],
    'import/no-dynamic-require': ['error'],
    'import/no-extraneous-dependencies': ['error', { packageDir: './' }],
    'import/extensions': ['error', {
      ts: 'never',
      tsx: 'never',
    }],

    'es/no-destructuring': 'error',

    '@typescript-eslint/explicit-member-accessibility': 'off',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/semi': 'error',
    '@typescript-eslint/member-delimiter-style': ['error', {
      multiline: {
        delimiter: 'semi',
        requireLast: true,
      },
      singleline: {
        delimiter: 'semi',
        requireLast: true,
      },
    }],
    '@typescript-eslint/consistent-type-assertions': ['error', {
      assertionStyle: 'as',
      objectLiteralTypeAssertions: 'allow-as-parameter',
    }],
  },
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {},
    },
  },
};

```

**Expected Result**
I would expect it throws error when found the above case, asking to change to the `as` casting format. 

**Actual Result**
It don't throw anything, seems like the code is fine.

**Additional Info**
The stranger behavior that I can't find anywhere the explanation is that when I run my project the rule throws on console the error. Not as I configured, but it throws. When I run `eslint .` it don't find any error.
I don't understand anything about this, BUT I believe it could be some conflict with `jsx` and the casting format.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.7.1` |
| `@typescript-eslint/parser`        | `3.7.1` |
| `TypeScript`                       | `3.7.2 |
| `ESLint`                           # I dont really know about this version since it's from the CRA
| `node`                             | `10.19.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
I am unable to repro this against master.

![image](https://user-images.githubusercontent.com/7462525/89565552-953b8a80-d7d3-11ea-8b4f-84c1b03f44d8.png)

Could you please provide a repro repo?

**Comment by Alecell** [comment]
Yeah I can, but what tool can I use to provide such a thing about `eslint`? Do you know one? At least I can share a zipped repro if that is allowed.

BTW, sorry if I misunderstood something but with `assertionStyle: 'as'` these below passed tests shouldn't be an error?
`const x = <Foo>new Generic<int>();`
`const x = <A>b`
`const x = <readonly number[]>[1];`
`const x = <a | b>('string');`

As them can be wrote with `as` format:
```
const x = new Generic<number>() as number;
const x = sim as number;
const x = [1] as readonly number[];
const x = ('string') as number | string;
```

**Comment by bradzacher** [comment]
>  Do you know one?

Create a git repo? Create a gist?

**Comment by Alecell** [comment]
Hey @bradzacher I was creating a repo so I realize that was my mistake after all.

Thanks for your help!
